### DISCUSSION TITLE
Using Nuxt3 programmatically

### DISCUSSION THREAD

**Discussion by Schetnan** [nuxt/nuxt#17845] [Questions]
In Nuxt 2, we had the ability to initialize Nuxt programmatically so that we could wrap it up into a corporate express Instance.  I am wondering how to do the same with Nuxt 3?  The code looked something like the following:

`async function configureNuxt(app) {

  const config = require('../nuxt.config.js')
  const isDev = config.env.delivery == 'development'
  const nuxt = await loadNuxt(isDev ? 'dev' : 'start')
  app.use(nuxt.render)

  if (isDev) {
    build(nuxt)
  }

  return nuxt
}`

**Comment by ruslankonev** [comment]
Now you need to do this → [Node.js function](https://v3.nuxtjs.org/docs/deployment/presets/node)

**Reply by Schetnan** [reply]
Thank you!

**Reply by Schetnan** [reply]
This seems fine for a deployment. What about dev? Is there a way to run in DEV hosted by an express server? We have an express library that provides authentication etc. up front to the request and we need the authentication credentials to be propagated to the request. We were able to do this pretty simply in both development and production mode in nuxt2. Is there a way with nuxt3 to supply the Express server to the development environment?

**Reply by danielroe** [reply]
You could (instead of using nuxt programmatically) add your express server in `~/server/middleware`, which will let it run on every request...

**Comment by wobsoriano** [comment]
As @danielroe  said, you can add an express server in `server/middleware` for example

```ts
// server/middleware/express.ts

import express from 'express'

const app = express()

app.use(express.json())
app.use(express.urlencoded({ extended: true}))

app.get('/authentication', (req, res) => {
    return res.json({
        auth: true
    })
})

export default app
```

Then you should be able to access `http://localhost:3000/authentication`

**Reply by Schetnan** [reply]
Thank you.  The problem is a little bit different than adding an endpoint for authentication.  Authentication happens through a corporate mechanism.  There is an authentication (express) middleware NPM library that resolves the authentication token and provides it to the request.  The library is counting on an 'express' API which means it uses APIs on the request etc - for example things like req.query.  In Nuxt2, nuxt was another piece of middleware that fit into the express chain.  We could register the authentication library, then initialize nuxt later.  By the time the request reached the Nuxt app, the authentication principal was already on the request.  Now, it seems that a project called H3 has control of the framework which is sort of express light, but not really express - it uses the bare bones node request rather than the express request.  The nuxt apps seems to be a compiled javascript file rather than a piece of middleware to register.  The original poster clued me into a preset that can be used to tie in Nuxt in product to an express server.  However, I am trying to figure out if there is a way to do this as part of the development stack so that it fits in with our corporate authentication technology.  I hope this describes the problem a little better?  Basically, I'd like to do something like the preset strategy for production builds - fire up an express stack so that I can add in all the corporate middleware first - then insert Nuxt into the stack so that it has all the corporate resources available and running  by the time it hits the nuxt app.

**Reply by danielroe** [reply]
No, you can't do this out of the box with Nuxt dev. You could likely create a specialised dev preset that would do what you wanted.

See:
*  https://github.com/nuxt/framework/blob/9d40a271ee5c65494a14b5ca8690cd194d1723c6/packages/nitro/src/runtime/entries/dev.ts
* https://github.com/nuxt/framework/blob/1e69bc56bb3efbb1b1ab6001006f8f96efaf2622/packages/nitro/src/server/dev.ts
* https://github.com/nuxt/framework/blob/8d6b97a4ac8d74ede1ce135215b791126b3e9c90/packages/nitro/src/presets/dev.ts

Or, you could fully initialise and export an express app within the middleware - it should work as you expect...

**Reply by Schetnan** [reply]
@danielroe   Interesting.. Gotta admit, not quite sure what I am looking at in the three files.  Looks interesting though.  It would be great to have an understanding of how to create a custom preset.  Regarding option #2.  I am having some trouble visualizing how that would work.

1). How do I get the pieces of server middleware to execute only once setting up an express server?
2). Event if I did, I'm not clear how that would solve my problem.  Something to play around with and experiment with though.

**Comment by JohnMica** [comment]
it would be interesting to know how this change  would impact dev and performance when Nuxt is simply an addition to the stack as @Schetnan mentioned.

**Comment by glopezep** [comment]
Is there an example of a custom preset using something different for example fastify?

**Comment by hnviradiya** [comment]


**Reply by danielroe** [reply]
I don't believe that is the case. You are welcome to open an issue with a reproduction.

**Comment by JoinGz** [comment]
I encountered the same problem today, is it not possible for nuxt3 to support running programmatically?

**Reply by danielroe** [reply]
Nuxt 3 does support running programmatically.

**Reply by kiplangatbryan** [reply]
Hey @danielroe I am facing the same issue here. I am trying to migrate my code base from nuxt2 to nuxt3, the problem was i am using nuxt2 programmatically but i can't seem to make it work i tried the link provided by @ruslankonev but it seems it does not work. Could you provide an example of using nuxt3 programmatically.

**Reply by hnviradiya** [reply]
Yes. Please. This is the only thing I am unable to upgrade to nuxt3 and for that matter vue3. I'm stuck as well.

**Reply by danielroe** [reply]
This is too general to help with. What specifically are you trying to do programmatically?

The link above is now found at https://nuxt.com/docs/getting-started/deployment#nodejs-server.

**Reply by hnviradiya** [reply]
Following is what I am trying to upgrade Nuxt 3. But don't see any equivalent method for that.

https://github.com/ColonelBundy/nuxtjs-nestjs-starter/blob/fbed6583717b42677d450ce570eb23102ecbe657/server/nuxt/index.ts#L16

This above instance, then we are able to use as filter to serve nuxt from nestjs
[server.useGlobalFilters(new NuxtFilter(nuxt));](https://github.com/ColonelBundy/nuxtjs-nestjs-starter/blob/fbed6583717b42677d450ce570eb23102ecbe657/server/main.ts#LL18C3-L18C49)

**Reply by danielroe** [reply]
The equivalent can be seen here:

https://github.com/nuxt/nuxt/blob/29f29302cecebf210c24f26f7b9a304f9299f4f7/packages/nuxi/src/commands/build.ts#L16-L59

**Reply by hnviradiya** [reply]
@danielroe Thanks a lot. I got very close to upgrade. What is equivalent of nuxt.render in Nuxt 3 (Following is full code)?

https://github.com/ColonelBundy/nuxtjs-nestjs-starter/blob/fbed6583717b42677d450ce570eb23102ecbe657/server/nuxt/nuxt.filter.ts#LL25C9-L25C42

**Reply by hnviradiya** [reply]
Phew. I am tired now figuring this out. Now I am no longer going to use nuxt at all. Will seperate out both server and client and will use only vue instead of monorepo using nuxt. If something is supposed to be easy but not making things easy why am I using an extra layer.

**Reply by danielroe** [reply]
Sorry to hear that - and sorry I missed your comment above.

I wish you the best, and don't hesitate to ask if you need any help in future.

**Reply by hnviradiya** [reply]
Sorry, it was said out of frustration. I still need it. Unfortunately ChatGPT is also not updated one to help on this issue. Can you please still help me on this one. 

I just need to serve nuxt app from NestJs, this way I am able to control nuxt more on where to allow nuxt app and not. 

It's been 3 years (even before nuxt 3 was stable release and was in beta stage) that I am struggling to serve Nuxt 3 with NestJs. I have tried numerous way you suggested. It seems even code which you have provided depends on many other things from Nuxt code which I believe is not ideal for me and also not working as per my previous comment in this thread. 

I believe some minimal example somewhere would be great. It's not only me who is doing things this way but probably others have given up on this.

**Reply by chriscalo** [reply]
I'm having this exact same problem. Please advise what is the identical drop-in replacement for Nuxt 2's `nuxt.render(req, res, next)` function. This feels like a regression from Nuxt 2, but maybe it's there in Nuxt 3 and just not documented.

**Reply by danielroe** [reply]
It's absolutely possible via a custom Nitro preset, and there a lot of documentation here in https://nitro.unjs.io. See this for example: https://github.com/unjs/nitro/blob/main/src/runtime/entries/node-server.ts.

Maybe if you share some code, it might be easier to get a specific response.

**Reply by chriscalo** [reply]
Here's a minimal example repo showing what we're trying to accomplish:

https://github.com/chriscalo/nuxt2-render-example

I'll repeat the key idea from the README here:

```js
const express = require("express");
const nuxtServer = require("./nuxt");
const server = express();

server.enable("trust proxy");
server.set("json spaces", 2);

// This pattern of integrating Nuxt's serving logic into our server gives us
// full control over the serving logic in all the ways we want, instead of us
// having to integrate our middleware into some server that we don't control.

// It's not hard to imagine all kinds of middleware we need to run before and
// after Nuxt:
server.use(require("./security"));
server.use(require("./logging"));
server.use(require("./parsing"));
server.use(require("./auth"));
server.use(require("./sessions"));

// Finally, apply Nuxt handlers:
server.use(nuxtServer);

const PORT = 3000;
server.listen(PORT, () => {
  console.log(`App listening on port ${PORT}`);
});
```

Of course there are other ways to accomplish this, but what's important to
realize is there are many reasons why developers either can't or don't want to
integrate their middleware into Nuxt's server and instead need to integrate
Nuxt's serving logic as middleware into their own server that they control.

**Reply by chriscalo** [reply]
Here is an example of doing the same thing with Astro. It felt refreshingly easy to get this working, similar to Nuxt 2.

This file is our main server, which imports a thin wrapper around the Astro serving logic.

```js
// server/index.js
import express from "express";
import authn, { signinRequired } from "./authn.js";
import api from "./api.js";
import astro from "./astro.js";

const server = express();

server.use(authn);

server.use([
  "/home",
  "/settings",
  // etc
], signinRequired);

server.use("/api", api);

server.use(astro);

server.listen(8080, function () {
  console.log("Server started on http://localhost:8080/");
});
```

And this file encapsulates importing and invoking the Astro SSR handler. It also calls Astro's `build()` function in development (just one way of doing this). Most importantly, this is all clearly described in the Astro documentation: [Node integration](https://docs.astro.build/en/guides/integrations-guide/node/), [`build()` function](https://docs.astro.build/en/reference/cli-reference/#build).

```js
// server/astro.js
import { resolve, dirname } from "node:path";
import { fileURLToPath } from "node:url";
import express from "express";
import { build } from "astro";

const DEVELOPMENT = (process.env.NODE_ENV !== "production");
const __dirname = dirname(fileURLToPath(import.meta.url));

const server = express();
export default server;

const astroHandlers = startAstro();

async function startAstro() {
  if (DEVELOPMENT) {
    await build({
      root: resolve(__dirname, "../"),
    });
  }
  
  const staticPath = resolve(__dirname, "../dist/client/");
  
  return {
    staticHandler: express.static(staticPath),
    ssrHandler: (await import("~/dist/server/entry.mjs")).handler,
  };
}

// serve Astro's static assets
server.use(async function (req, res, next) {
  const { staticHandler } = await astroHandlers;
  staticHandler(req, res, next);
});

// serve Astro's SSR pages and endpoints
server.use(async function (req, res, next) {
  const { ssrHandler } = await astroHandlers;
  const { locals } = res;
  ssrHandler(req, res, next, locals);
});
```

Because this was so easy to get working with Astro, I'm probably going to continue down that path, so my need here is resolved. But I wanted to still describe roughly what I think everyone else here is asking for in case that might lead to a workable solution.

**Comment by brooklynb7** [comment]
I have the same issue and same requirement.
Just want to use nuxt.render in the express/koa programmatically.
Just want Nuxt3 as a middleware in Express/Koa, but not Express/Koa as a middleware in Nuxt3.

**Comment by msyfls123** [comment]
I succeeded running Nuxt inside Express by following steps:

## Step 1
Modify nuxt.config.ts with [Nitro Config](https://nitro.unjs.io/deploy/runtimes/node#handler-advanced).
```ts
export default defineNuxtConfig({
  nitro: {
    preset: 'node',
  },
})
```

## Step 2
Run `nuxt build`.

You will got such a line in `.output/server/index.mjs`.
```js
export { F as handler, D as listener, E as websocket } from './chunks/runtime.mjs';
```

## Step 3

Write an index file like blew, and install `express` as you need.

```js title=index.js
import express from 'express';

import { listener } from './.output/server/index.mjs';

const app = express();

app.get('/hello', (req, res) => {
  res.send('Hello, World!');
});

app.use(express.static('./.output/public'));
app.use(listener);

app.listen(8080, '0.0.0.0', () => {
  console.log('Server listening on http://0.0.0.0:8080');
})
```

That's it!

**Reply by ghost** [reply]
@msyfls123  Does it working with server side render?
